File: sample-projects/go/domain/model/hit_points.go
package model

const MAX_HIT_POINTS = 999

// Hit points drop when the creature takes damage and recover when it rests in its lair.
type HitPoints struct {
	Current   int
	Max       int
	Temporary int
}

func InitHitPoints(max int) HitPoints {
	if max > MAX_HIT_POINTS {
		max = MAX_HIT_POINTS
	}
	return HitPoints{Current: max, Max: max, Temporary: 0}
}

func (hp HitPoints) TakeDamage(damage int) HitPoints {
	hp.Current -= damage
	if hp.Current < 0 {
		hp.Current = 0
	}
	return hp
}

func (hp HitPoints) Rest() HitPoints {
	hp.Current = hp.Max
	return hp
}

File: sample-projects/ruby/lib/de/sots/cellars_and_centaurs/domain/model/hit_points.rb
# frozen_string_literal: true

require_relative '../../version'

module De::Sots::CellarsAndCentaurs::Domain::Model
  # Hit points drop when the creature takes damage and recover when it rests in its lair.
  class HitPoints
    MAX_HIT_POINTS = 999

    attr_reader :current, :max, :temporary

    def initialize(current, max, temporary = 0)
      @current = current
      @max = [max, MAX_HIT_POINTS].min
      @temporary = temporary
    end

    def self.init(max)
      new(max, max, 0)
    end

    def take_damage(amount)
      HitPoints.new([current - amount, 0].max, max, temporary)
    end

    def rest
      HitPoints.new(max, max, 0)
    end

    def alive?
      current.positive?
    end
  end
end
